embed-code-go/fragmentation/fragmentation.go

```diff
@@ -0,0 +1,245 @@
+	"fmt"
+	"os"
+	"path/filepath"
+)
+
+const (
+	FragmentStart = "#docfragment"
+	FragmentEnd   = "#enddocfragment"
+)
+
+// Splits the given file into fragments and writes them into corresponding output files.
+//
+// SourcesRoot a full path of the root directory of the source code to be embedded.
+//
+// CodeFile a full path of a file to fragment.
+type Fragmentation struct {
+	Configuration configuration.Configuration
+	SourcesRoot   string
+	CodeFile      string
+}
+
+//
+// Initializers
+//
+
+// Builds Fragmentation from the given codeFileRelative and config.
+//
+// codeFileRelative is a relative path to a code file to fragment.
+func NewFragmentation(
+	codeFileRelative string,
+	config configuration.Configuration,
+) Fragmentation {
+
+	fragmentation := Fragmentation{}
+
+	sourcesRootRelative := config.CodeRoot
+
+	absoluteSourcesRoot, err := filepath.Abs(sourcesRootRelative)
+	fragmentation.SourcesRoot = absoluteSourcesRoot
+	if err != nil {
+		panic(err)
+	}
+
+	absoluteCodeFile, err := filepath.Abs(codeFileRelative)
+	fragmentation.CodeFile = absoluteCodeFile
+	if err != nil {
+		panic(err)
+	}
+
+	fragmentation.Configuration = config
+
+	return fragmentation
+}
+
+//
+// Public methods
+//
+
+// Splits the file into fragments.
+//
+// Returns a refined content of the file to be cut into fragments, and the Fragments.
+func (fragmentation Fragmentation) Fragmentize() ([]string, map[string]Fragment, error) {
+	fragmentBuilders := make(map[string]*FragmentBuilder)
+	var contentToRender []string
+
+	file, err := os.Open(fragmentation.CodeFile)
+	if err != nil {
+		panic(err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text() + "\n"
+		contentToRender, fragmentBuilders, err = fragmentation.parseLine(line, contentToRender, fragmentBuilders)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	fragments := make(map[string]Fragment)
+	for k, v := range fragmentBuilders {
+		fragments[k] = v.Build()
+	}
+	fragments[DefaultFragment] = CreateDefaultFragment()
+
+	return contentToRender, fragments, nil
+
+}
+
+// Serializes fragments to the output directory.
+//
+// Keeps the original directory structure relative to the sources root dir.
+// That is, `SRC/src/main` becomes `OUT/src/main`.
+func (fragmentation Fragmentation) WriteFragments() error {
+	allLines, fragments, err := fragmentation.Fragmentize()
+	if err != nil {
+		return err
+	}
+
+	EnsureDirExists(fragmentation.targetDirectory())
+
+	for _, fragment := range fragments {
+		fragmentFile := NewFragmentFileFromAbsolute(fragmentation.CodeFile, fragment.Name, fragmentation.Configuration)
+		fragment.WriteTo(fragmentFile, allLines, fragmentation.Configuration)
+	}
+	return nil
+}
+
+//
+// Static functions
+//
+
+// Searches for code files with patterns defined in configuration
+// and fragmentizes them with creating fragmented files as a result.
+func WriteFragmentFiles(configuration configuration.Configuration) error {
```

Comment:
Improve this doc now it starts with `Searches ....` while the method name is `Write ...` This may be confusing. With the first sentence, I would expect what we **write** here, and details of **how** we **write** should go to the following sentences.